### DISCUSSION TITLE
Current Data vs Data

### DISCUSSION THREAD

**Discussion by abhieshekumar** [reduxjs/redux-toolkit#2861] [Q&A] [answered]
I have been trying to understand the difference between currentData and data -

1. Is it that data will have the freshest copy while current data holds on the data provided in last successful fetch.
2. Also, if the later half is true does it also imply the current data is undefined only if the query has never resolved successfully.
3. I was trying to log currentData [here](https://codesandbox.io/s/rtk-query-loading-state-5blb9d?file=/src/App.js), but couldn't see it. Please let me know if I have missed anything?

**Comment by markerikson** [comment, marked as answer]
> Is it that data will have the freshest copy while current data holds on the data provided in last successful fetch.

Yep, this is it!

> I was trying to log currentData [here](https://codesandbox.io/s/rtk-query-loading-state-5blb9d?file=/src/App.js), but couldn't see it.

Looks like the sandbox is using RTK 1.6, and we added that in either 1.7 or 1.8.  Bump the RTK version and it should be there.

**Reply by josias-r** [reply]
@markerikson This is the JSDoc of `currentData`:
```
Where `data` tries to hold data as much as possible, also re-using data from the last arguments passed into the hook, this property will always contain the received data from the query, for the current query arguments.
```
This seems to imply exactly the opposite of `Is it that data will have the freshest copy while current data holds on the data provided in last successful fetch.`
Because `currentData` will **not** hold on to data, but **always contain the received data**.

While recently starting to work with RTK Query, these two distincion can't seem to get into my head, and I have to look it up everytime 😅

**Reply by phryneas** [reply]
@josias-r 

You start with `usePostsQuery({ page: 3 })` - `data` will be for page 3 and `currentData` will be for page 3.
You change to `usePostsQuery({ page: 4 })` - while page 4 is not loaded yet, `data` will stay page 3 so you have something to display, while `currentData` gives you the data for page 4 - `undefined`, as nothing is fetched yet.
Data is loaded for page 4: `data` will be data for page 4 and `currentData` will be data for page 4.

**Reply by josias-r** [reply]
@phryneas Yup thank you! I keep checking this in the code and finding out the same result, but with your comment I can come back to this every time thread from now on haha.

I think it slowly gets into my head as `current` as a term indicates to be more _fleeting_. So `current` will be more temporary and reset to `undefined` when the query changes and it is still fetching

**Reply by phryneas** [reply]
`currentData` is always bound exactly to the argument that is passed in, while data is trying to always have *some* value that you can display. 
Most apps will probably only use `data`, in combination with some loading indicator or graying UI items out.
Does that make sense?

**Reply by wkigenyi** [reply]
Yes it makes sense.

**Reply by jtiret** [reply]
I've observed another distinction between `data` and `currentData` and am curious whether this behavior is intentional or a bug. Previously, I relied on data rather than `currentData` and frequently used `api.util.updateQueryData` for optimistic updates. I've noticed that when a query is in progress (e.g., triggered by polling), updates made via `api.util.updateQueryData` are not reflected in data, but are visible in `currentData` (even when I haven't altered the hook argument at all). This discrepancy isn't apparent when queries are fast, but with queries taking 200-300ms and a short polling interval, optimistic updates happening within this timeframe fail to apply.

**Reply by nostrapollo** [reply]
> I've observed another distinction between `data` and `currentData` and am curious whether this behavior is intentional or a bug. Previously, I relied on data rather than `currentData` and frequently used `api.util.updateQueryData` for optimistic updates. I've noticed that when a query is in progress (e.g., triggered by polling), updates made via `api.util.updateQueryData` are not reflected in data, but are visible in `currentData` (even when I haven't altered the hook argument at all). This discrepancy isn't apparent when queries are fast, but with queries taking 200-300ms and a short polling interval, optimistic updates happening within this timeframe fail to apply.

@phryneas I'm looking at the same situation as explained in this comment above, before I create another issue/discussion I wanted to check if this is intended design or if there is a pattern to have `data` reflect changes to it's cache value before the request is completed. Let me know if this is better suited to a new issue, or more detail is required.

**Reply by phryneas** [reply]
I guess it should reflect those changes immediately. PRs and investigation would be very welcome.

**Comment by jukart** [comment]
I just found the following behaviour as problematic:
- we have a successful query result, which results in data and currentData to be the same and matching the query parameters
- now the query parameter is changed to skipToken, which results in data to stay as it was before and currentData change to undefined

I my opinion data should also be updated to undefined because this is the result of the current query (no result with skipToken).

**Reply by EskiMojo14** [reply]
`data` is the last *successful* result, `undefined` is not a successful result. Skipping doesn't necessarily mean you don't want to display the old data - and if it does, you have `currentData` and the status flags to decide that for yourself.
